### PR TITLE
✨ Bring download/upload progress logs to the frontend

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
Allow to redirect tqdm library progress logs to the frontend to show the user when:
- downloading workspace
- downloading/uploading ports

Bonus:
- use tqdm contrib functions to make logging behave better when a progress bar is displayed.
NOTE: no progress shown for archiving/unarchiving since async/sync code is proving difficult to circumvent.

[Screencast from 25. 08. 22 17:04:15.webm](https://user-images.githubusercontent.com/35365065/186707581-bf3db869-69d3-4b17-a988-33c92634903e.webm)

## Related issue/s

improves user experience https://github.com/ITISFoundation/osparc-issues/issues/638

### CODE REVIEW

File: packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py [hunk 1]
```diff
@@ -61,11 +62,23 @@ async def _file_chunk_reader(
...
+    file: Path,
+    response: ClientResponse,
+    pbar: tqdm,
+    io_log_redirect_cb: Optional[LogRedirectCB],
+):
     async with aiofiles.open(file, "wb") as file_pointer:
         while chunk := await response.content.read(CHUNK_SIZE):
             await file_pointer.write(chunk)
-            pbar.update(len(chunk))
+            if pbar.update(len(chunk)) and io_log_redirect_cb:
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py:80]: Q: so ``pbar`` always prints in stdout right? regardless of ``io_log_redirect_cb``?

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port_utils.py [hunk 2]
```diff
@@ -168,6 +169,7 @@ async def pull_file_from_store(
     key: str,
     fileToKeyMap: Optional[dict[str, str]],
     value: FileLink,
+    io_log_redirect_cb: Optional[LogRedirectCB],
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port_utils.py:172]: why did not you choose to have ``None`` by default?

**Review by pcrespov — APPROVED** [review]
Very nice! 
How do you think rabbit will react to so many messages? 
Progress is not critical information. Should we somehow downsample the steps at risk of loosing some progress steps?

File: packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py (hunk 1 above)
Comment (sanderegg) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py:80]: ok, so there are 2 things. pbar.update returns True if it should actually print (cause it does not always print based on some internal rationale like minimum amount of iterations, time, etc etc...).
Then you can also get all the progress info as a dictionary. or as how it prints it by itself.
Also I think that tqdm prints to stderr

Comment (mrnicegyu11) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py:80]: fyi on this topic: RabbitMQ is thought to be able to handle approx. <50k msg/sec

Comment (sanderegg) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_common/file_io_utils.py:80]: concerning the througput. these are logs. treated as logs are. I guess nobody ever checked if rabbitmq is under/over used. something to chec at some point.
and I tend to disagree with saying that progress is not critical. this is user feedback and is very important.

File: packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port_utils.py (hunk 2 above)
Comment (sanderegg) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port_utils.py:172]: hmm yes. so I tend to put the default None to the external interface, not the internals of the library. I think defaults are usually a bit dangerous as people tend to not set them explicitly.

**Review by mguidon — APPROVED** [review]
+1 for the video

**Pull request merged**